### DISCUSSION TITLE
Need assistance for using for_each

### DISCUSSION THREAD

**Discussion by anvithpotluri** [redpanda-data/connect#2981] [General]
part of my code is below. I have grabbed an array of 100 values and want to iterate and add them to my timestamp and values arrays. I get the error: "expected array value" when I try this code. Am I doing something wrong with the foreach section?
    
  
    - mapping: |
        root.value_array = json("valueArray")

        # If value_array is not an array, wrap it in one
        if root.value_array.type() != "array" {
          root.value_array = [ root.value_array ]
        }
    - for_each:
        for_each: root.value_array
        processors:
          - bloblang: |
              # Calculate the timestamp for each value
              timestamp = start_ts + (10 * index)
              root.timestamps.append(timestamp)
              root.values.append(value)

**Comment by mihaitodor** [comment]
Hey @anvithpotluri 👋 I think it might be easier if you share some sample input and desired output. Otherwise, my best guess is that you just need to use bloblang. For example, if you have this input:

```json
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

Try this mapping:
```coffee
root.values = this
root.timestamps = range(0, this.length()).map_each(i -> i * 10)
```

Alternatively, The `fold()` method is a bit more versatile:

```coffee
root.values = this
root.timestamps = this.fold({"timestamps": [], "index": 0}, v -> {"timestamps": v.tally.timestamps.append(10 * v.tally.index), "index": v.tally.index + 1}).timestamps
```

**Reply by anvithpotluri** [reply]
I tried the following below, and it failed. How can I properly create my timestamp array? I need it to be the same length as the values arrray and want each value to be calculated with the formula start_ts + (10 * index)


        root.timestamps = json("valueArray").map_each(
          value -> start_ts + (10 * value.index()) )

**Reply by anvithpotluri** [reply]
@mihaitodor any idea how to get an array populated successfully in bloblang with the formula start_ts + (10 * index) ?

**Comment by anvithpotluri** [comment]
Here is my full Pipeline. the important fields from the input are metric, dataTS, and a valueArray. The first value in the valueArray needs to correspond to the dataTS and each value after corresponds to the dataTS + (index * 10). This is how I want to fill out my values and timestamps.
An example input would be: valueArray: [10, 11, 12, 13], dataTS: 1730317343630.
The output I would want is: values: [10,11,12,13], timestamps: [1730317343630, 1730317343640, 1730317343650, 1730317343660]

EDIT: writing this out makes me realize the lack of need of a for loop. All I would need to do is make an array with the length of the values, and then have this array filled out with timestamps with each being 10 greater than the last.

```


pipeline:
  processors:
    - bloblang: |
        root = if content().contains("NaN") { content().replace_all("NaN", "\"NaN\"").parse_json() }

    - mapping: |
        root.metric.__name__ = json("metric")

        # Extract the start timestamp
        start_ts = json("dataTS").number()
        # Initialize arrays for timestamps and values
        root.timestamps = []
        root.values = []
    - mapping: |
        root.value_array = json("valueArray")

        # If value_array is not an array, wrap it in one
        if root.value_array.type() != "array" {
          root.value_array = [ root.value_array ]
        }
    - for_each:
        for_each: root.value_array
        processors:
          - bloblang: |
              # Calculate the timestamp for each value
              timestamp = start_ts + (10 * index)
              root.timestamps.append(timestamp)
              root.values.append(value)

```

**Reply by mihaitodor** [reply]
Note that the syntax there is invalid: `timestamp = start_ts + (10 * index)`. You have to use `let` to declare a variable and then `$` when referencing it. Also, you can't declare a variable in one processor and then reference it in another one. You'd have to use metadata for that purpose.
